heidrun/nara_json.rb

```diff
@@ -0,0 +1,575 @@
+#     <year>[VALUE]</year>
+#     <logicalDate>[VALUE]</logicalDate>
+#   </proposableQualifiableDate>
+# </releaseDateArray>
+#
+# A record can have both dc:date and dcterms:temporal values, like
+# coverageDates and broadcastDateArray in
+# https://catalog.archives.gov/api/v1?pretty=true&resultTypes=item%2CfileUnit&objects.object.@objectSortNum=1&naIds=5860128
+
+# Return a string suitable for sourceResource.date.providedLabel
+#
+def make_date_provided_label(date)
+  date_string(date.node)
+  # node = date.node['proposableQualifiableDate']
+  # node = [node] unless node.is_a? Array
+  # node.map { |n| date_string(n) }
+end
+
+# Return a string for sourceResource.temporal.begin
+#
+def make_begin_date(dates)
+  date_string(dates.node['coverageStartDate'])
+end
+
+# Return a string for sourceResource.temporal.end
+#
+def make_end_date(dates)
+  date_string(dates.node['coverageEndDate'])
+end
+
+Krikri::Mapper.define(:nara_json, :parser => Krikri::JsonParser) do
+  provider :class => DPLA::MAP::Agent do
+    uri 'http://dp.la/api/contributor/nara'
+    label 'National Archives and Records Administration'
+  end
+
+  dataProvider :class => DPLA::MAP::Agent do
+    providedLabel record.field('description', 'item | itemAv | fileUnit',
+                               'physicalOccurrenceArray',
+                               'itemPhysicalOccurrence |' \
+                                 'itemAvPhysicalOccurrence |' \
+                                 'fileUnitPhysicalOccurrence',
+                               'referenceUnitArray', 'referenceUnit',
+                               'name')
+  end
+
+  isShownAt :class => DPLA::MAP::WebResource do
+    uri record.field('naId').first_value.map { |id| nara_catalog_uri(id) }
+  end
+
+  object :class => DPLA::MAP::WebResource do
+    uri record.field('objects', 'object', 'file').first_value
+              .map { |o| make_obj_uri(o) }
+    dcformat record.field('objects', 'object', 'file').first_value
+                   .map { |o| make_obj_dcformat(o) }
+  end
+
+  preview :class => DPLA::MAP::WebResource do
+    uri record.field('objects', 'object', 'thumbnail').first_value
+              .map { |o| make_preview_uri(o) }
+    dcformat record.field('objects', 'object', 'thumbnail').first_value
+                   .map { |o| make_preview_dcformat(o) }
+  end
+
+  originalRecord :class => DPLA::MAP::WebResource do
+    uri record_uri
+  end
+
+  sourceResource :class => DPLA::MAP::SourceResource do
+
+    # <parentRecordGroup> OR <parentCollection>
+    #   <naId>[VALUE]</naId>
+    #   <title>[VALUE]</title>
+    #   <recordGroupNumber>[VALUE]</recordGroupNumber>
+    # </parentRecordGroup> OR </parentCollection>
+    collection :class => DPLA::MAP::Collection do
+      title record.field('description', 'item | itemAv | fileUnit',
+                         'parentSeries',
+                         'parentRecordGroup | parentCollection', 'title')
+    end
+
+    contributor :class => DPLA::MAP::Agent do
```

Comment:
Since `contributor` looks like it's likely to get an array, you probably want to use the `:each` argument at this level (e.g., see the [ESDN mapping lines 41-46](https://github.com/dpla/heidrun-mappings/blob/master/heidrun/esdn_mods.rb#L41-46).)